src/Voron/Impl/Journal/JournalReader.cs

```diff
@@ -671,19 +530,191 @@ void AssertValidLastPageNumber(TransactionHeader* transactionHeader)
             }
         }
 
-        private bool CanIgnoreDataIntegrityErrorBecauseTxWasSynced(TransactionHeader* currentTx, StorageEnvironmentOptions options)
+        public readonly HashSet<Guid> RecoveredJournalIds = [];
+        
+        private bool TryReadAndValidateHeader(StorageEnvironmentOptions options, ref Pager.PagerTransactionState txState, out TransactionHeader* current)
+        {
+            for (; _readAt4Kb < _journalPagerNumberOfAllocated4Kb; _readAt4Kb++)
+            {
+                if (TryValidateTransaction(options, ref txState, out current) is false)
+                {
+                    Debug.Assert(current != null, "current != null");
+
+                    if (VerifyNoUnexpectedValidTransactionsAfter(options, ref txState))
+                    {
+                        // we found a _valid_ transaction (and all the invalid ones were
+                        // already synced, so we can safely ignore them), so we'll go 
+                        // forward from there
+                        _readAt4Kb--;
+                        continue;
+                    }
+
+                    return false;
+                }
+
+                RecoveredJournalIds.Add(current->JournalId);
+
+                _next4Kb = _readAt4Kb + GetTransactionSizeIn4Kb(current);
+
+                if (JournalId == Guid.Empty)
+                {
+                    JournalId = current->JournalId;
+                }
+
+                if (current->Flags == TransactionPersistenceModeFlags.LinkedJournalsRecord &&
+                    _environment.Options.RootJournal is null) // this only applies to the _root_, not to branches
+                {
+                    ProcessLinkedJournalsRecord(current);
+                    _readAt4Kb += GetTransactionSizeIn4Kb(current) - 1;
+                    continue;
+                }
+                
+                if ((current->JournalId == JournalId) is false &&
+                    current->JournalId != Guid.Empty) // this may be legacy
+                {
+                    // not our env, skip processing it
+                    _readAt4Kb += GetTransactionSizeIn4Kb(current) - 1;
+                    continue;
+                }
+
+                if (current->JournalId == Guid.Empty)
+                {
+                    if (Legacy_IsOldTransactionFromRecycledJournal(current))
+                    {
+                        _readAt4Kb += GetTransactionSizeIn4Kb(current) - 1;
+                        continue;
+                    }
+                    // Here we are dealing with a valid transaction (in terms of tx id)
+                    // that has zeroed DatabaseId, probably a legacy transaction for the 
+                    // current database (non-shared journal mode), allowing it
```

Comment:
How the legacy dbs are going to work? After recovering they will start using shared journal mode automatically, right?

Do we ensure that the legacy journals will be removed during immediately after the recovery? So the next attempt to do the recovery will not attempt to operate in legacy mode